proposals/0002-gep.md

```diff
@@ -0,0 +1,225 @@
+---
+MEP: 0002
+Title: Unifed API
+Discussion: Link
+Implementation: Link
+---
+
+# Unified API
+
+## Abstract
+
+This GEP discusses the unified API that handles the interaction with LLM providers. This represents the feature set for MVP.
+
+## Motivation
+
+
+The Unified API abstracts much of the overhead of interacting with LLM providers.
+It stores all the provider specific logic in a single place and allowing developers to focus on the core logic. It also introduces durability to LLM apps. Additionally, unifying the API allows optimizations to be leveraged such as latency and cost improvements.
+
+
+### Requirements
+
+- R1: User should be able to specify specific models or select auto mode (selection handled by Glide)
+- R2: Should be a Hertz Client
+- R3: Should take a user request as input
+- R4: Handles all provider specific logic
+- R5: Easily maintained
+- R6: API schemas must unify common LLM request params (e.g. topP, N, etc)
+- R7: API routes must unify common LLM endpoints/API
+- R8: API schemas should allow to customize provider-specific params (e.g. a different messages for OpenAI/Anthropic)
+- R9: Focus on building out a single endpoint for /Chat. Most apps will be using this and then we can focus our efforts on making this part of our API super effective
+- R10: Disallow overrides in the MVP. All params are updated via config. This will change over time but I want users to tell which params they want to override (and how) instead of us making the assumption.
+
+
+## Design
+
+
+- A Hertz client that handles an incoming request and returns a reponse from that provider.
+- Provider specific clients
+
+### Routes
+```yaml
+routes:
+    chat: /v1/pools/{pool-id}/chat/
+    complete: /v1/pools/{pool-id}/completions/ # future route, not for MVP
+    multi-modal: /v1/pools/{pool-id}/multimodal/ # future route, not for MVP
+    embedding: /v1/pools/{pool-id}/embed/ # future route, not for MVP
+```
+
+
+#### User Request Schema for Chat Route
+
+Below is an example user request to the Chat route. The other routes will look very similar, if not identical. The provider specific params will look different.
+
+Example OpenAI Request Body:
+
+```json
+{
+  "message": [
+      {
+        "role": "system",
+        "content": "You are a helpful assistant."
+      },
+      {
+        "role": "user",
+        "content": "tell me a joke"
+      }
+    ],
+  "messageHistory": ["Hello there", "How are you?", "I'm good, how about you?"]
```

Comment:
@mkrueger12 hm, wondering why OpenAI `messageHistory` is not annotated by the role? 🤔

Reply:
I'm not sure but I updated it to reflect this: https://platform.openai.com/docs/guides/text-generation/chat-completions-api